python/pyarrow/tests/test_convert_builtin.py

```diff
@@ -639,3 +639,10 @@ def test_structarray_from_arrays_coerce():
         pa.StructArray.from_arrays(arrays)
 
     assert result.equals(expected)
+
+
+def test_decimal_array_with_none_and_nan():
+    values = [decimal.Decimal('1.234'), None, np.nan, decimal.Decimal('nan')]
+    array = pa.array(values)
+    assert array.type == pa.decimal128(4, 3)
```

Comment:
can you add a test here with an explicit decimal type sufficient to accommodate the data?